solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java

```diff
@@ -1806,9 +1806,11 @@ public void waitForState(
     AtomicReference<DocCollection> docCollection = new AtomicReference<>();
     CollectionStateWatcher watcher =
         (n, c) -> {
-          docCollection.set(c);
           boolean matches = predicate.matches(n, c);
-          if (matches) latch.countDown();
+          if (matches) {
+            docCollection.set(c);
```

Comment:
We may need to keep this outside? what if condition doens't match and timeout happens below. We are accessing collection below
```
throw new TimeoutException(
            "Timeout waiting to see state for collection="
                + collection
                + " :"
                + docCollection.get());
```